src/AsyncDicomReader.js

```diff
@@ -0,0 +1,412 @@
+import { ReadBufferStream } from "./BufferStream";
+import { ValueRepresentation } from "./ValueRepresentation";
+import {
+    EXPLICIT_BIG_ENDIAN,
+    EXPLICIT_LITTLE_ENDIAN,
+    IMPLICIT_LITTLE_ENDIAN,
+    VM_DELIMITER,
+    UNDEFINED_LENGTH,
+    TagHex
+} from "./constants/dicom";
+import { Tag } from "./Tag";
+import { DicomMessage, singleVRs } from "./DicomMessage";
+import { DicomMetaDictionary } from "./DicomMetaDictionary";
+import { DicomMetadataListener } from "./utilities";
+
+/**
+ * This is an asynchronous binary DICOM reader.
+ *
+ * Assume this is still preliminary as to the exact interface as it is
+ * initially being released for testing purposes only.
+ *
+ * There is no support for compressed streams.
+ */
+export class AsyncDicomReader {
+    syntax = EXPLICIT_LITTLE_ENDIAN;
+
+    constructor(options = {}) {
+        this.isLittleEndian = options?.isLittleEndian;
+        this.stream = new ReadBufferStream(null, this.isLittleEndian, {
+            clearBuffers: true,
+            ...options
+        });
+    }
+
+    /**
+     * Reads the preamble and checks for the DICM marker.
+     * Returns true if found/read, leaving the stream past the
+     * marker, or false, having not found the marker.
+     */
+    async readPreamble() {
+        const { stream } = this;
+        await stream.ensureAvailable();
+        stream.reset();
+        stream.increment(128);
+        if (stream.readAsciiString(4) !== "DICM") {
+            stream.reset();
+            return false;
+        }
+        return true;
+    }
+
+    async readFile(options = undefined) {
+        const hasPreamble = await this.readPreamble();
+        if (!hasPreamble) {
+            throw new Error("TODO - can't handle no preamble file");
+        }
+        this.fmi = await this.readFmi(options);
+        const listener = options?.listener || new DicomMetadataListener();
+        this.dict ||= {};
+        listener.startObject(this.dict);
+        await this.read(listener, options);
+        this.dict = listener.pop();
+
+        return this;
+    }
+
+    async readFmi(options = null) {
```

Comment:
Not clear what Fmi is here.  Can you use a more descriptive symbol to make the code more self-documenting?